internal/querynode/query_shard.go

```diff
@@ -223,14 +223,16 @@ func (q *queryShard) getNewTSafe(tp tsType) (Timestamp, error) {
 }
 
 func (q *queryShard) waitUntilServiceable(ctx context.Context, guaranteeTs Timestamp, tp tsType) {
+	st := q.getServiceableTime(tp)
+	log.Debug("wait serviceable ts check", zap.String("tsType", tp.String()), zap.Uint64("guarantee ts", guaranteeTs), zap.Uint64("serviceable ts", st), zap.String("channel", q.channel))
 	q.watcherCond.L.Lock()
 	defer q.watcherCond.L.Unlock()
```

Comment:
getServiceableTime shall be protected by `watchCond.L.Lock`, otherwise, in some extreme case, last ts change will be missed.

Reply:
Modified. Now we can ensure the serviceableTime is retrieved after the lock acquisition. Please check it again.